### PR TITLE
GPIO refactor (Simplify macro, reduce type states, and support interrupt)

### LINKED ISSUE #83: External interrupts (example, question)
I wrote a simple example of GPIO-driven interrupt handling, available [here.](https://github.com/jkboyce/stm32f3xx-hal/blob/master/examples/gpio_interrupt.rs) Compile with `cargo build --example gpio_interrupt --features "rt stm32f303xc"` On the F3 Discovery board it toggles an led when you push the user button.

I'm curious if anyone's working on–or has thoughts about–adding better support for interrupts. The [`gpio.rs`](https://github.com/stm32-rs/stm32f4xx-hal/blob/master/src/gpio.rs) in `stm32f4xx-hal` has some nice methods based on the `ExtiPin` trait, making it easy to set up interrupts without direct register access. For example:

```rust
let gpioc = dp.GPIOC.split();
let mut board_btn = gpioc.pc13.into_floating_input();
board_btn.make_interrupt_source(&mut dp.SYSCFG);
board_btn.enable_interrupt(&mut dp.EXTI);
board_btn.trigger_on_edge(&mut dp.EXTI, Edge::RISING);
```

It would be nice to get the f3/f4 HALs into closer alignment but I wanted to gauge others' thoughts before jumping in.

### PR DESCRIPTION
**Pull request by Piroro-hs**
* Simplify macro
  `gpio.rs` uses more generics than macro now
* Collapse Input type arguments
  An attempt to address #125
* Support interrupt
  Close #83 and #112

### CODE REVIEW

**Comment by Sh3Rm4n** [comment]
Super interesting attempt. I'll take a closer look in the next few days. But this approach seems promising!

File: src/gpio.rs [hunk 1]
```diff
@@ -50,1205 +46,1289 @@ pub trait GpioExt {
     fn split(self, ahb: &mut AHB) -> Self::Parts;
 }
 
-/// Input mode (type state)
-pub struct Input<MODE> {
-    _mode: PhantomData<MODE>,
+/// GPIO Register interface traits private to this module
```
Comment (Sh3Rm4n) [inline src/gpio.rs:49]: Also known as sealed traits :)

File: src/gpio.rs [hunk 2]
```diff
@@ -50,1205 +46,1289 @@ pub trait GpioExt {
...
+impl Index for Ux {
+    fn index(&self) -> u8 {
+        self.0
+    }
+}
 
-/// Alternate function 9 (type state)
-pub struct AF9;
+/// Input mode (type state)
+pub struct Input;
```
Comment (Sh3Rm4n) [inline src/gpio.rs:160]: Maybe add a documentation notice, that the Input type is not encoded in a type state anymore?
The user has to know, that it is their responsibility to set the correct input type and that this 
is due issues explained in https://github.com/stm32-rs/stm32f3xx-hal/issues/125. 

The change notice belonging to the changelog, the overall behavior should be written down here.

File: src/gpio.rs [hunk 3]
```diff
@@ -50,1205 +46,1289 @@ pub trait GpioExt {
...
+/// Output mode (type state)
+pub struct Output<OTYPE>(PhantomData<OTYPE>);
```
Comment (Sh3Rm4n) [inline src/gpio.rs:162]: The elements of tuples are always public, iirc. Is this intended or a problem?

File: src/gpio.rs [hunk 4]
```diff
@@ -50,1205 +46,1289 @@ pub trait GpioExt {
...
+/// Alternate function (type state)
+pub struct Alternate<AF, OTYPE>(PhantomData<AF>, PhantomData<OTYPE>);
```
Comment (Sh3Rm4n) [inline src/gpio.rs:164]: I've noticed many changes, where `Alternate<PushPull>` is now enforced. 
Is this really needed for every case? [`stm32f4xx-hal`](https://github.com/stm32-rs/stm32f4xx-hal/blob/dadf2192e6ade2652de3b02754c3606532dd624d/src/gpio.rs#L34-L42) does only differentiate between `Alternate` and `Alternate<OpenDrain>` / `AlternateOD`. 

Is there a use case for a neutral `Alternate` Type State, where it does not matter, which Output State is used?
Point 3. in https://github.com/stm32-rs/stm32f3xx-hal/issues/125#issue-665855804 might be a problem, right? 

What is your thought about this?

**Review by Sh3Rm4n — COMMENTED** [review]
Some thoughts which came into my mind. 

I've only read a small part of `src.gpio.rs` so far. Will look into this in more detail later.

File: src/gpio.rs (hunk 4 above)
Comment (Piroro-hs) [reply src/gpio.rs:164]: I follow the [stm32f1xx-hal](https://github.com/stm32-rs/stm32f1xx-hal/blob/ca01418e356d2879628dd86cc4c63597d85c1665/src/gpio.rs#L484-L530) as it feels more consistent.

Some peripherals should accept `Alternate<MODE>` instead of `Alternate<PushPull>`, maybe?

File: src/gpio.rs (hunk 3 above)
Comment (Piroro-hs) [reply src/gpio.rs:162]: Tuple elements are [treated like normal struct fields](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=966728d4d6f3661263b58275910d4483), so this is ok 😃

File: src/gpio.rs (hunk 2 above)
Comment (Piroro-hs) [reply src/gpio.rs:160]: Yeah, documentation update is needed.

However, I currently think about re-adding `into_input_xxx()` methods which implemented like
```rust
pub fn into_input_floating(self, moder: &mut GPIO::MODER, pupdr: &mut GPIO::PUPDR) -> Pin<GPIO, INDEX, Input> {
    pupdr.floating(self.index.index());
    self.into_input(moder)
}
```

**Comment by David-OConnor** [comment]
Could you please add support for GPIO interrupts? I'm actually modding this now to do so. I made a standalone module a while back for this (In my PR), but am switching it to GPIO to be consistent with L4 and F4 HALs.

**Comment by Piroro-hs** [comment]
GPIO interrupts support added.
Referred to f1xx-hal, f4xx-hal, and #200.

**Comment by Piroro-hs** [comment]
Clippy warnings should be fixed by #202

File: src/gpio.rs (hunk 4 above)
Comment (Piroro-hs) [reply src/gpio.rs:197]: Support for OD pins is added for serial and pwm peripherals.

File: src/gpio.rs [hunk 5]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+/// at runtime, and erases the specific compile time type of the GPIO.
+/// It does only matter, that it is a GPIO pin with a specific MODE.
+///
+/// See [examples/gpio_erased.rs] as an example.
+///
+/// [examples/gpio_erased.rs]: https://github.com/stm32-rs/stm32f3xx-hal/blob/v0.6.0/examples/gpio_erased.rs
+pub type PXx<MODE> = Pin<Gpiox, Ux, MODE>;
+
+macro_rules! modify_at {
+    ($xr:expr, $bits:expr, $i:expr, $val:expr) => {
```
Comment (seanybaggins) [inline src/gpio.rs:263]: `pin_num` could be a more informative name than `i`.

Comment (seanybaggins) [reply src/gpio.rs:263]: Perhaps adding required tokens to the macro could help improve readability when the macro is invoked.
Ex.
```rust
macro_rules! modify_at {
    (register: $xr:expr, some_name_1: $bits:expr, extri_num: $i:expr, some_name: $val:expr) => {
        ...
    }    
}
```

File: src/gpio.rs [hunk 6]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+    GPIO: Gpio,
+    INDEX: Index,
+    MODE: Active,
+{
+    /// Make corresponding EXTI line sensitive to this pin
+    pub fn make_interrupt_source(&mut self, syscfg: &mut SysCfg) {
+        let i = self.index.index() % 4;
+        let extigpionr = self.gpio.port_index() as u32;
+        match self.index.index() {
+            0..=3 => unsafe { modify_at!(syscfg.exticr1, 4, i, extigpionr) },
```
Comment (seanybaggins) [inline src/gpio.rs:514]: Consider adding some let expression before the `modify_at` macro that labels what the number 4 is.

File: src/gpio.rs [hunk 7]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+            12..=15 => unsafe { modify_at!(syscfg.exticr4, 4, i, extigpionr) },
+            _ => unreachable!(),
+        }
+    }
+
+    /// Generate interrupt on rising edge, falling edge, or both
+    pub fn trigger_on_edge(&mut self, exti: &mut EXTI, edge: Edge) {
+        cfg_if! {
+            // if #[cfg(any(feature = "stm32f373", feature = "stm32f378"))] { // TODO: update after https://github.com/stm32-rs/stm32-rs/pull/496 released
+            if #[cfg(not(any(feature = "stm32f302", feature = "stm32f303", feature = "stm32f334")))] {
```
Comment (seanybaggins) [inline src/gpio.rs:526]: I am not sure this conditional compilation is correct. Looking at the [stm32-rs/STM32F301](https://stm32-rs.github.io/stm32-rs/STM32F301.html#EXTI:RTSR1). It does not fall in this feature category but does have a numbered RTSR.

File: src/gpio.rs [hunk 8]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+                let ftsr = &exti.ftsr1;
+            }
+        }
+        let (rise, fall) = match edge {
+            Edge::Rising => (true, false),
+            Edge::Falling => (false, true),
+            Edge::RisingFalling => (true, true),
+        };
+        unsafe {
+            modify_at!(rtsr, 1, self.index.index(), rise as u32);
```
Comment (seanybaggins) [inline src/gpio.rs:540]: Consider adding a let expression that labels what 1 is.

File: src/gpio.rs [hunk 9]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+            modify_at!(ftsr, 1, self.index.index(), fall as u32);
+        }
+    }
+
+    /// Enable external interrupts from this pin.
+    pub fn enable_interrupt(&mut self, exti: &mut EXTI) {
+        cfg_if! {
+            // if #[cfg(any(feature = "stm32f373", feature = "stm32f378"))] { // TODO: update after https://github.com/stm32-rs/stm32-rs/pull/496 released
+            if #[cfg(not(any(feature = "stm32f302", feature = "stm32f303", feature = "stm32f334")))] {
```
Comment (seanybaggins) [inline src/gpio.rs:549]: Same comment about correct conditional compilation. https://stm32-rs.github.io/stm32-rs/STM32F301.html#EXTI:IMR1

File: src/gpio.rs [hunk 10]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+            }
+        }
+        unsafe { modify_at!(imr, 1, self.index.index(), 1) };
+    }
+
+    /// Disable external interrupts from this pin
+    pub fn disable_interrupt(&mut self, exti: &mut EXTI) {
+        cfg_if! {
+            // if #[cfg(any(feature = "stm32f373", feature = "stm32f378"))] { // TODO: update after https://github.com/stm32-rs/stm32-rs/pull/496 released
+            if #[cfg(not(any(feature = "stm32f302", feature = "stm32f303", feature = "stm32f334")))] {
```
Comment (seanybaggins) [inline src/gpio.rs:562]: Same comment about correct conditional compilation. https://stm32-rs.github.io/stm32-rs/STM32F301.html#EXTI:IMR1

File: src/gpio.rs [hunk 11]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+            }
+        }
+        unsafe { modify_at!(imr, 1, self.index.index(), 0) };
+    }
+
+    /// Clear the interrupt pending bit for this pin
+    pub fn clear_interrupt_pending_bit(&mut self) {
+        cfg_if! {
+            // if #[cfg(any(feature = "stm32f373", feature = "stm32f378"))] { // TODO: update after https://github.com/stm32-rs/stm32-rs/pull/496 released
+            if #[cfg(not(any(feature = "stm32f302", feature = "stm32f303", feature = "stm32f334")))] {
```
Comment (seanybaggins) [inline src/gpio.rs:575]: Correct conditional compilation? https://stm32-rs.github.io/stm32-rs/STM32F301.html#EXTI:PR1

File: src/gpio.rs [hunk 12]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+            $(
+                af!($i, [<U $i>], [<AF $i>], [<IntoAf $i>],  [<into_af $i _push_pull>],  [<into_af $i _open_drain>]);
+            )+
+        }
+    };
+}
 
-        #[cfg(feature = "unproven")]
-        impl <MODE> toggleable::Default for PXx<Output<MODE>> {}
+af!([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
```
Comment (seanybaggins) [inline src/gpio.rs:656]: Wow that is a lot cleaner

File: src/gpio.rs [hunk 13]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
 
-                        fn set_low(&mut self) -> Result<(), Self::Error> {
-                            // NOTE(unsafe, write) atomic write to a stateless register
-                            unsafe { (*$GPIOX::ptr()).bsrr.write(|w| w.bits(1 << (16 + self.i))) }
-                            Ok(())
-                        }
+                impl Afr for AFRL {
+                    #[inline]
+                    fn afx(&mut self, i: u8, x: u8) {
+                        unsafe { modify_at!((*$GPIOX::ptr()).afrl, 4, i, x as u32); }
```
Comment (seanybaggins) [inline src/gpio.rs:841]: Consider let expression that labels what 4 is.

File: src/gpio.rs (hunk 7 above)
Comment (Piroro-hs) [reply src/gpio.rs:526]: The [stm32-rs/STM32F301](https://stm32-rs.github.io/stm32-rs/STM32F301.html#EXTI:RTSR1) is based on latest nightly version of stm32-rs, while we are using v0.12 for now.

File: src/gpio.rs [hunk 14]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
-    _mode: PhantomData<MODE>,
+use private::{Afr, GpioRegExt, Moder, Ospeedr, Otyper, Pupdr};
+
+/// Marker trait for GPIO ports
+pub trait Gpio: private::Gpio {}
+
+/// Marker trait for compile time defined GPIO ports
+pub trait GpioStatic: Gpio {
+    /// Associated MODER register
+    type MODER: Moder;
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/gpio.rs:129]: `Upper` is a trait bound for a type alias `UPPER`.

File: src/gpio.rs [hunk 15]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
-/// Output mode (type state)
-pub struct Output<MODE> {
-    _mode: PhantomData<MODE>,
+use private::{Afr, GpioRegExt, Moder, Ospeedr, Otyper, Pupdr};
+
+/// Marker trait for GPIO ports
+pub trait Gpio: private::Gpio {}
+
+/// Marker trait for compile time defined GPIO ports
+pub trait GpioStatic: Gpio {
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/gpio.rs:127]: Yup. https://doc.rust-lang.org/rust-by-example/trait/supertraits.html

**Comment by Sh3Rm4n** [comment]
> I'll take a closer look in the next few days. But this approach seems promising!

Next few months rather. I have this on my TODO list. I'm hopeful that I find time this week :)

File: src/gpio.rs (hunk 13 above)
Comment (Sh3Rm4n) [reply src/gpio.rs:841]: The `modify_at` macro could need some explanation, but 4 is essentially the `bitwidth` of the field in the 32 bit register, which is changed by this invocation.

File: src/gpio.rs [hunk 16]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+///
+/// See [examples/gpio_erased.rs] as an example.
+///
+/// [examples/gpio_erased.rs]: https://github.com/stm32-rs/stm32f3xx-hal/blob/v0.6.0/examples/gpio_erased.rs
+pub type PXx<MODE> = Pin<Gpiox, Ux, MODE>;
+
+macro_rules! modify_at {
+    ($xr:expr, $bits:expr, $i:expr, $val:expr) => {
+        $xr.modify(|r, w| {
+            w.bits(r.bits() & !(u32::MAX >> 32 - $bits << $bits * $i) | $val << $bits * $i)
```
[the remark this answers is not in the document]
Comment (Sh3Rm4n) [reply src/gpio.rs:265]: I had to take some time to understand it, but i'd suggest to write the macro like this and comment it a bit more, so that it is easier to parse. Without the knowledge of the operator evaluation order, this is hard to parse.

```suggestion
            let mask = !(u32::MAX >> (32 - $bits) << ($bits * $i));
            let value = $val << $bits * $i;
            w.bits(r.bits() & mask | value)
```

**Review by Sh3Rm4n — CHANGES_REQUESTED** [review]
Wow, this PR is huge. :eyes:  Both in code change, but also implementation wise. 

Props to you, to unclutter the macro and leveraging the rust type-system and static polymorphism techniques, to not rely too heavily on macros to hold it all together. 

Your extensive use of traits not only move most of the dispatch logic out of the macro, but actually makes it easier to read through the gpio implementation itself. 

The interrupt implementation is a very welcomed additions as well. 

Props to you  and many thanks for this implementation. :pray: 

I've added mostly nits as review comments, but design wise, this looks perfect. I'm not able to find any critiques, as I couldn't have done it any better than you. As if I could even do this myself :laughing:

File: src/gpio.rs [hunk 17]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
 
-/// Alternate function 0 (type state)
-pub struct AF0;
+/// Marker trait for readable pin modes
+pub trait Readable {}
 
-/// Alternate function 1 (type state)
-pub struct AF1;
+/// Marker trait for slew rate configurable pin modes
+pub trait OutputSpeed {}
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/gpio.rs:158]: Users can't implement these marker traits to pin mode typestates thanks to Rust's trait orphan rules.

File: src/gpio.rs [hunk 18]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
-pub struct Floating;
-/// Pulled down input (type state)
-pub struct PullDown;
-/// Pulled up input (type state)
-pub struct PullUp;
-
-/// Output mode (type state)
-pub struct Output<MODE> {
-    _mode: PhantomData<MODE>,
+use private::{Afr, GpioRegExt, Moder, Ospeedr, Otyper, Pupdr};
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/gpio.rs:121]: To use functions defined in these traits, we need this use statement :(

File: src/gpio.rs [hunk 19]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+                use super::{
+                    Input, Output, Analog, PushPull, OpenDrain,
+                    IntoAf0, IntoAf1, IntoAf2, IntoAf3, IntoAf4, IntoAf5, IntoAf6, IntoAf7,
+                    IntoAf8, IntoAf9, IntoAf10, IntoAf11, IntoAf12, IntoAf13, IntoAf14, IntoAf15,
+                    AF0, AF1, AF2, AF3, AF4, AF5, AF6, AF7, AF8, AF9, AF10, AF11, AF12, AF13, AF14, AF15,
+                };
+
+                #[allow(unused_imports)]
+                use typenum::{
+                    U0, U1, U2, U3, U4, U5, U6, U7, U8, U9, U10, U11, U12, U13, U14, U15
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/gpio.rs:776]: Used to implement partially erased pin type aliases.
```rust
pub type $PXi<MODE> = Pin<$Gpiox, $Ui, MODE>;
```

File: src/serial.rs [hunk 20]
```diff
@@ -46,42 +48,44 @@ pub unsafe trait TxPin<USART> {}
 /// RX pin - DO NOT IMPLEMENT THIS TRAIT
 pub unsafe trait RxPin<USART> {}
 
-unsafe impl TxPin<USART1> for gpioa::PA9<AF7> {}
-unsafe impl TxPin<USART1> for gpiob::PB6<AF7> {}
-unsafe impl TxPin<USART1> for gpioc::PC4<AF7> {}
-unsafe impl RxPin<USART1> for gpioa::PA10<AF7> {}
-unsafe impl RxPin<USART1> for gpiob::PB7<AF7> {}
-unsafe impl RxPin<USART1> for gpioc::PC5<AF7> {}
+unsafe impl<OTYPE> TxPin<USART1> for gpioa::PA9<AF7<OTYPE>> {}
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/serial.rs:51]: Allowing open-drain configuration makes it possible to use the interface on high voltage bus'es using a simple pull-up, so we should keep it. It is also required for USART half-duplex mode (currently not supported in hal, though).

File: src/gpio.rs [hunk 21]
```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
...
+            w.bits(r.bits() & !(u32::MAX >> 32 - $bits << $bits * $i) | $val << $bits * $i)
+        });
+    };
+}
 
+impl<GPIO, INDEX, MODE> Pin<GPIO, INDEX, MODE>
+where
+    INDEX: Unsigned,
+{
+    /// Erases the pin number from the type
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/gpio.rs:274]: Is it ok to use single letter `impl<G, I, M>`? `Gpio` and `Index` will collide with trait names 😥

[the remark this answers is not in the document]
Comment (Sh3Rm4n) [reply src/gpio.rs:274]: Hm. That is a good point. I don't really know the best solution. Maybe a namespace aka modules would help? I would prefer changing the name(space) of the traits though. So we can distinguish trait `trait::Gpio` vs. type `type:Gpio`? 

Do you have better naming ideas?

File: src/serial.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Sh3Rm4n) [reply src/serial.rs:51]: Good point. And good thinking in the future. Let's keep it as is.

File: src/gpio.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sh3Rm4n) [reply src/gpio.rs:526]: Updated to stm32f3 v0.13 in  #211

File: examples/gpio_interrupts.rs [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    let mut user_button = gpioa
+        .pa0
+        .into_pull_down_input(&mut gpioa.moder, &mut gpioa.pupdr);
+    user_button.make_interrupt_source(&mut syscfg);
+    user_button.trigger_on_edge(&mut exti, Edge::Rising);
+    user_button.enable_interrupt(&mut exti);
+    // Moving ownership to the global BUTTON so we can clear the interrupt pending bit.
+    cortex_m::interrupt::free(|cs| *BUTTON.borrow(cs).borrow_mut() = Some(user_button));
+
+    unsafe { NVIC::unmask(Interrupt::EXTI0) };
```
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply examples/gpio_interrupts.rs]: `nvic(&self) -> pac::Interrupt` method is added.

File: src/gpio.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Piroro-hs) [reply src/gpio.rs]: Naming collision resolved by moving marker traits under `marker` module :)

**Review by Sh3Rm4n — APPROVED** [review]
Minor tiny nit. 

Overall this looks fantastic. Let me know if you want me to press the merge button :grinning:

**Comment by Piroro-hs** [comment]
Updated (add comments and move `IntoAf..` traits to `marker` module).
This is ready to merge :)

**Comment by Sh3Rm4n** [comment]
Nice. Let's merge it :steam_locomotive: 

Thanks for your hard work, again! :)
